design/aws-ecs-scheduled-ecs-task-construct.md

```diff
@@ -0,0 +1,141 @@
+# AWS ECS - L3 Construct for Scheduling EC2 Tasks
+
+To address issue [#2352](https://github.com/awslabs/aws-cdk/issues/2352), the ECS CDK construct library should provide a way for customers to create a standalone scheduled task, without creating a service, for their container. The task will be initiated by a cloudwatch event that is scheduled based on a [CW Schedule Event](http://docs.aws.amazon.com/AmazonCloudWatch/latest/events/ScheduledEvents.html).
+
+This would mean adding a new ECS CDK construct `ScheduledEc2Task` and properties `ScheduledEc2TaskProps`, that would take in the necessary properties required to create a Task Definition, an EventRuleTarget as well as an AWS EventRule. To allow for flexibility, container overrides will also be supported properties.
+
+Note: Currently, there is only [CloudFormation](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-reference-ecs.html) support for scheduled tasks on EC2, and not on Fargate. CloudFormation currently only supports clusters, services and taskdefinitions: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-reference-ecs.html.
+
+TaskDefinitions contain a network mode, and for EC2, we can use none, bridge or the many others. To support Fargate, we would need to use a awsvpc as the NetworkMode of the TaskDefinition, but this requires us to create a NetworkConfiguration, which is not supported in the TaskDefinition, but is supported in the Service.
+
+For this construct, we're creating a scheduled task (that is standalone task, not backed with a service), and for Fargate to support this feature, we'd need to have CloudFormation create the concept of a Standalone Task/RunTask (not Task Definition) that can take a NetworkConfiguration as a property. It works for ECS because we're using bridge as the NetworkMode, which does not require a NetworkConfiguration to be present.
+
+## General approach
+
+The new [`ecs.ScheduledEc2Task`] class will include an L3 construct for:
+
+* ScheduledEc2Task
+
+A `ScheduledEc2Task` will create a task definition with the specified container. An `Ec2EventRuleTarget` will be created and any container overrides will be applied to an AWS `EventRule` (indicating how frequently the task should be run). Based on the `EventRule` schedule, a task will run on the EC2 instances specified in the cluster. 
+
+The following container overrides can be passed to the container to enable flexibility:
```

Comment:
How are these container overrides wired up?  IIRC no container overrides are supported by CloudWatch Events, it only uses what's in the task definition.  From below it doesn't look like it's really overrides, it's just container settings?

Reply:
It's passed in as inputOptions to addTarget for the event rule.

```ts
    rule.addTarget(target, {
      jsonTemplate: JSON.stringify({
        containerOverrides: [{
          name: 'ScheduledTaskContainer',
          environment: props.environment
        }]
      })
    });
```

Reply:
Ohhh I see.  I think that's probably more useful if you are attempting to re-use a task definition for multiple rules.  For this construct, it seems to be 1:1 mapping between task definition and rule, so probably doesn't provide as much convenience.

Reply:
It's a fair point, but I would think of a scheduled task as being something that I might want to run adhoc for instance if it runs scheduled daily, but I wanted to run the same task with different environment variables or resource constraints (cpu, etc). 

Based on this tweet from Nathan: https://twitter.com/nathankpeck/status/1120719301349126145 felt like it would be something our customers would benefit from.

Reply:
For adhoc requests, you're calling RunTask directly, in which case you would specify the overrides in the API call.

If you're creating multiple scheduled rules that run the same task, but differ by environment variables, I can definitely see that use case.  However, since it's infrastructure as code + CDK, it's pretty cheap to create 2 ScheduledEc2Task objects, each with the same image but different env variables.  I would expect each ScheduledEc2Task objects to create its own task definition with the different env variables.

Reply:
Sounds good!